ledger/eval/eval.go

```diff
@@ -1590,22 +1598,33 @@ func isAbsent(totalOnlineStake basics.MicroAlgos, acctStake basics.MicroAlgos, l
 	return lastSeen+basics.Round(allowableLag) < current
 }
 
-func (eval *BlockEvaluator) activeChallenge() challenge {
-	current := eval.Round()
-	var round basics.Round
-	if current > challengeInterval {
-		lastChallenge := current - (current % challengeInterval)
-		// challengeRound is in effect if we're after one grace period, but before the 2nd ends.
-		if current > lastChallenge+challengeGracePeriod && current < lastChallenge+2*challengeGracePeriod {
-			round = lastChallenge
-			challengeHdr, err := eval.state.BlockHdr(round)
-			if err != nil {
-				panic(err)
-			}
-			return challenge{round, challengeHdr.Seed, challengeBits}
-		}
+type headerSource interface {
+	BlockHdr(round basics.Round) (bookkeeping.BlockHeader, error)
+}
+
+func activeChallenge(proto *config.ConsensusParams, current uint64, headers headerSource) challenge {
+	rules := proto.Mining()
+	// are challenges active?
+	if rules.ChallengeInterval == 0 || current < rules.ChallengeInterval {
+		return challenge{}
+	}
+	lastChallenge := current - (current % rules.ChallengeInterval)
+	// challenge is in effect if we're after one grace period, but before the 2nd ends.
+	if current <= lastChallenge+rules.ChallengeGracePeriod ||
+		current > lastChallenge+2*rules.ChallengeGracePeriod {
+		return challenge{}
+	}
+	round := basics.Round(lastChallenge)
+	challengeHdr, err := headers.BlockHdr(round)
+	if err != nil {
+		panic(err)
+	}
+	challengeProto := config.Consensus[challengeHdr.CurrentProtocol]
+	// challenge is not considered if rules have changed since that round
```

Comment:
If rules have changed, what is the expected behavior - extended grace period?

Reply:
That challenge just gets ignored.  No big deal, we do them a lot.